src/transformStateWithClones.js

```diff
@@ -7,7 +7,26 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    } else if (action.type === 'clear') {
+      currentState = {};
+    }
```

Comment:
[CODE STYLE]: Use a switch statement instead of multiple if/else if conditions when you have a limited amount of conditions to make the code more readable and maintainable.